### ISSUE TITLE
PerfDiff: EtlDiffer never sets its regression verdict, so the presence of ETL traces downgrades any real BenchmarkDotNet regression to "noise" (exit 0)

### ISSUE THREAD

**Issue by rjmurillo** [rjmurillo/moq.analyzers#1265]
## Summary

An internal code audit found that `EtlDiffer.TryCompareETL` computes and prints an ETL overweight report but **never assigns its `regression` out-parameter** — it is initialized to `false` and returned as `false` on every successful comparison. `PerfDiff.CompareAsync` treats a successful ETL comparison with `etlRegressionDetected == false` as proof that a BenchmarkDotNet-detected regression was "noise" and returns exit code **0**, even when `--failOnRegression` is set.

This matters because PerfDiff's exit code is a required PR merge gate: `build/scripts/perf/ComparePerfResults.ps1` (line 25) runs PerfDiff with `--failOnRegression`, and per `docs/architecture/ADR-008-benchmarkdotnet-perfdiff-for-performance-regression-detection.md` that exit code decides whether the PR merges. The defect is latent today only because Linux CI produces no ETW traces — the first time `*.etl.zip` artifacts appear in both folders, the perf gate is silently disarmed for every regression.

This is a specific correctness defect in the tool, distinct from the open PerfDiff test-coverage and refactoring issues (#594–#601, #609–#613).

Two smaller defects in the same file are folded in because their fixes touch the same members:

- The catch filter at `EtlDiffer.cs:31` omits `System.IO.InvalidDataException`, so a corrupt `*.etl.zip` (ETL traces are zip archives; `TraceLog.OpenOrConvert` unzips them) crashes PerfDiff instead of degrading gracefully.
- The overweight math has unguarded edge cases: `baselineTotal == 0` produces an infinite `growth` (`EtlDiffer.cs:126`), and `expectedDelta == 0` produces an `Infinity` or `NaN` `overweight` (`EtlDiffer.cs:138`, `:140`), which poisons the `Interest` score and the report sort.
- `EtlDiffer` writes to `Console.Error`/`Console.WriteLine` (`:28`, `:33`) instead of the `ILogger` abstraction used everywhere else in PerfDiff.

## Current behavior

`src/tools/PerfDiff/ETL/EtlDiffer.cs:13-36`:

```csharp
    public static bool TryCompareETL(string sourceEtlPath, string baselineEtlPath, out bool regression)
    {
        regression = false;

        try
        {
            using TraceLog sourceTraceLog = TraceLog.OpenOrConvert(sourceEtlPath)
                ?? throw new InvalidOperationException($"Failed to open ETL trace file: {sourceEtlPath}");
            using TraceLog baselineTraceLog = TraceLog.OpenOrConvert(baselineEtlPath)
                ?? throw new InvalidOperationException($"Failed to open ETL trace file: {baselineEtlPath}");

            CallTree sourceCallTree = GetCallTree(sourceTraceLog, sourceEtlPath);
            CallTree baselineCallTree = GetCallTree(baselineTraceLog, baselineEtlPath);
            ImmutableArray<OverWeightResult> report = GenerateOverweightReport(sourceCallTree, baselineCallTree);

            Console.WriteLine(string.Join(Environment.NewLine, report.Take(10)));
            return true;
        }
        catch (Exception ex) when (ex is InvalidOperationException or IOException or UnauthorizedAccessException)
        {
            Console.Error.WriteLine($"ETL comparison failed: {ex.Message}");
            return false;
        }
    }
```

`regression` is assigned exactly once (`regression = false;` at line 15) and never again — the success path (lines 17–29) computes `report` but derives no verdict from it.

The consumer, `src/tools/PerfDiff/PerfDiff.cs:33-47`:

```csharp
        (bool etlCompareSucceeded, bool etlRegressionDetected) = CheckEltTraces(baselineFolder, resultsFolder, logger);
        if (!etlCompareSucceeded)
        {
            logger.LogWarning("We detected a regression in BenchmarkDotNet and there is no ETL info.");
            return failOnRegression ? 1 : 0;
        }

        if (etlRegressionDetected)
        {
            logger.LogWarning("We detected a regression in BenchmarkDotNet and there _is_ ETL info which agrees there was a regression.");
            return failOnRegression ? 1 : 0;
        }

        logger.LogTrace("We detected a regression in BenchmarkDotNet but examining the ETL trace determined that it was noise.");
        return 0;
```

**Concrete input → wrong exit code:**

1. `--baseline` and `--results` folders each contain one `*.etl.zip` (any pair of valid traces) alongside BDN `*full-compressed.json` files.
2. The results run has a genuine 4x regression; `BenchmarkDotNetDiffer` detects it (`regressionDetected == true`).
3. `TryCompareETL` opens both traces, prints the report, returns `true` with `regression == false` — unconditionally.
4. `etlRegressionDetected == false` routes to the "noise" branch → **`return 0`** despite `--failOnRegression`. The regression merges.

Additional wrong outcomes in the same code:

- **Corrupt zip:** a truncated/corrupt `results/*.etl.zip` makes `TraceLog.OpenOrConvert` throw `InvalidDataException` (zip decompression), which the catch filter at line 31 does not cover → unhandled exception, exit 1 with a raw stack trace (an unactionable false block, instead of the intended graceful `(false, …)` degradation).
- **Degenerate totals:** `LoadTrace` sums only `ExclusiveMetric` into `total` (`EtlDiffer.cs:117`) while `data` accumulates `InclusiveMetric`, so `baselineTotal` can legitimately be `0` with non-empty data → `growth = sourceTotal / baselineTotal` is `+Infinity` (`:126`); `expectedDelta == 0` (e.g. `growth` rounding to exactly `1f`) makes `overweight = delta / expectedDelta * 100` `±Infinity` or `NaN` (`:138`, `:140`), corrupting `Interest` scoring (`:143`) and the sort. Cosmetic today; load-bearing the moment any verdict is derived from the report.

## Desired behavior

ETL analysis must never convert a BenchmarkDotNet-detected regression into exit 0 while it produces no verdict of its own.

**Recommended fix (implemented below): remove the ETL veto.** Treat the ETL overweight report as advisory diagnostics only: when BDN detects a regression, PerfDiff exits `failOnRegression ? 1 : 0` regardless of ETL availability, and the ETL report (if traces exist) is logged to help diagnose the regression. `TryCompareETL` loses its dead `out bool regression` parameter so the type system stops implying a verdict that is never produced.

**Alternative considered (not recommended now):** derive a verdict from the report using the code's own constants, e.g. `regression = report.Any(static r => r.Interest >= 3 && Math.Abs(r.Overweight) > 110);` — `|Overweight| > 110` is the report's own overweight-signal cut (`EtlDiffer.cs:143`) and `Interest >= 3` requires at least two corroborating signals (share of delta and stack position, `:144-148`). This is not wired in because CI currently produces no ETL traces to calibrate against; enabling an uncalibrated verdict would re-arm the exact veto being removed (false pass) or create false blocks. If ETL verdicts are wanted later, they should ship behind the report-derived expression above plus calibration data.

Also: catch `InvalidDataException`; guard the overweight math; route all output through `ILogger` (the only caller, `PerfDiff.CheckEltTraces` at `PerfDiff.cs:66`, already has a logger to pass).

## Implementation plan

1. **`src/tools/PerfDiff/PerfDiff.cs`** — replace `CompareAsync` and `CheckEltTraces` (the latter renamed to reflect its new advisory role). Complete replacement members:

```csharp
    internal static async Task<int> CompareAsync(
        string baselineFolder, string resultsFolder, bool failOnRegression, ILogger logger, CancellationToken token)
    {
        token.ThrowIfCancellationRequested();

        (bool compareSucceeded, bool regressionDetected) = await BenchmarkDotNetDiffer.TryCompareBenchmarkDotNetResultsAsync(baselineFolder, resultsFolder, logger, token).ConfigureAwait(false);

        if (!compareSucceeded)
        {
            logger.LogError("Failed to compare the performance results see log.");
            return 1;
        }

        if (!regressionDetected)
        {
            logger.LogTrace("No performance regression found.");
            return 0;
        }

        // The ETL overweight report is advisory-only: EtlDiffer does not produce a regression
        // verdict, so it must never override the BenchmarkDotNet result. See the linked issue.
        if (TryReportEtlDiagnostics(baselineFolder, resultsFolder, logger))
        {
            logger.LogWarning("We detected a regression in BenchmarkDotNet; see the ETL overweight report above for the likely hot paths.");
        }
        else
        {
            logger.LogWarning("We detected a regression in BenchmarkDotNet and there is no ETL info.");
        }

        return failOnRegression ? 1 : 0;
    }
```

```csharp
    private static bool TryReportEtlDiagnostics(string baselineFolder, string resultsFolder, ILogger logger)
    {
        // try look for ETL traces
        if (!TryGetETLPaths(baselineFolder, logger, out string? baselineEtlPath))
        {
            return false;
        }

        if (!TryGetETLPaths(resultsFolder, logger, out string? resultsEtlPath))
        {
            return false;
        }

        return EtlDiffer.TryCompareETL(resultsEtlPath, baselineEtlPath, logger);
    }
```

   (Delete the old `CheckEltTraces`; `TryGetETLPaths` and `ETLFileExtension` are unchanged. `CompareAsync` sits inside the existing `#pragma warning disable CA1848/CA2254` region — keep it.)

2. **`src/tools/PerfDiff/ETL/EtlDiffer.cs`** — replace `TryCompareETL`: drop the `out` parameter, take `ILogger`, add `InvalidDataException` to the filter, log through the logger. Add `using Microsoft.Extensions.Logging;` to the file's usings (`InvalidDataException` is in `System.IO`, already covered by implicit usings):

```csharp
    /// <summary>
    /// Compares two ETL traces and logs an overweight report for diagnostics.
    /// This comparison is advisory only: it does not produce a regression verdict,
    /// so callers must not use its result to override a BenchmarkDotNet verdict.
    /// </summary>
    /// <param name="sourceEtlPath">Path to the ETL trace from the run under test.</param>
    /// <param name="baselineEtlPath">Path to the baseline ETL trace.</param>
    /// <param name="logger">Logger for the report and errors.</param>
    /// <returns><see langword="true"/> if both traces were opened and the report was produced; otherwise <see langword="false"/>.</returns>
    public static bool TryCompareETL(string sourceEtlPath, string baselineEtlPath, ILogger logger)
    {
        try
        {
            using TraceLog sourceTraceLog = TraceLog.OpenOrConvert(sourceEtlPath)
                ?? throw new InvalidOperationException($"Failed to open ETL trace file: {sourceEtlPath}");
            using TraceLog baselineTraceLog = TraceLog.OpenOrConvert(baselineEtlPath)
                ?? throw new InvalidOperationException($"Failed to open ETL trace file: {baselineEtlPath}");

            CallTree sourceCallTree = GetCallTree(sourceTraceLog, sourceEtlPath);
            CallTree baselineCallTree = GetCallTree(baselineTraceLog, baselineEtlPath);
            ImmutableArray<OverWeightResult> report = GenerateOverweightReport(sourceCallTree, baselineCallTree, logger);

#pragma warning disable CA1848, CA2254 // LoggerMessage delegates; template built from report content
            logger.LogInformation(string.Join(Environment.NewLine, report.Take(10)));
#pragma warning restore CA1848, CA2254
            return true;
        }
        catch (Exception ex) when (ex is InvalidOperationException or IOException or InvalidDataException or UnauthorizedAccessException)
        {
#pragma warning disable CA1848 // LoggerMessage delegates
            logger.LogError(ex, "ETL comparison failed: {Message}", ex.Message);
#pragma warning restore CA1848
            return false;
        }
    }
```

   The only caller to update is `PerfDiff.CheckEltTraces` (`src/tools/PerfDiff/PerfDiff.cs:66`) — already rewritten in step 1. No other call sites exist (verified by search).

3. **`src/tools/PerfDiff/ETL/EtlDiffer.cs`** — replace `GenerateOverweightReport`, promoting the `ComputeOverweights` local function to an `internal static` method (unit-testable: it takes only primitives and dictionaries) with the math guards. `LoadTrace` stays as-is (as a static local of `GenerateOverweightReport`):

```csharp
    private static ImmutableArray<OverWeightResult> GenerateOverweightReport(CallTree source, CallTree baseline, ILogger logger)
    {
        float sourceTotal = LoadTrace(source, out Dictionary<string, float> sourceData);
        float baselineTotal = LoadTrace(baseline, out Dictionary<string, float> baselineData);

        return ComputeOverweights(sourceTotal, sourceData, baselineTotal, baselineData, logger);

        static float LoadTrace(CallTree callTree, out Dictionary<string, float> data)
        {
            data = new Dictionary<string, float>();
            float total = 0;
            foreach (CallTreeNodeBase? node in callTree.ByID)
            {
                if (node.InclusiveMetric == 0)
                {
                    continue;
                }

                string key = node.Name;
                data.TryGetValue(key, out float weight);
                data[key] = weight + node.InclusiveMetric;

                total += node.ExclusiveMetric;
            }

            return total;
        }
    }

    internal static ImmutableArray<OverWeightResult> ComputeOverweights(
        float sourceTotal,
        Dictionary<string, float> sourceData,
        float baselineTotal,
        Dictionary<string, float> baselineData,
        ILogger logger)
    {
        if (sourceTotal <= 0 || baselineTotal <= 0)
        {
#pragma warning disable CA1848 // LoggerMessage delegates
            logger.LogWarning("Cannot compute ETL overweight report: source total is {SourceTotal} and baseline total is {BaselineTotal}; both must be positive.", sourceTotal, baselineTotal);
#pragma warning restore CA1848
            return ImmutableArray<OverWeightResult>.Empty;
        }

        if (sourceTotal == baselineTotal)
        {
            return ImmutableArray<OverWeightResult>.Empty;
        }

        float totalDelta = sourceTotal - baselineTotal;
        float growth = sourceTotal / baselineTotal;
        ImmutableArray<OverWeightResult>.Builder results = ImmutableArray.CreateBuilder<OverWeightResult>();
        foreach (string key in baselineData.Keys)
        {
            // skip symbols that are not in both traces
            if (!sourceData.ContainsKey(key))
            {
                continue;
            }

            float baselineValue = baselineData[key];
            float sourceValue = sourceData[key];
            float expectedDelta = baselineValue * (growth - 1);
            float delta = sourceValue - baselineValue;

            // Guard: expectedDelta == 0 (baselineValue == 0, or growth rounding to exactly 1)
            // makes 'delta / expectedDelta' Infinity or NaN, which poisons the Interest score
            // and the sort. Treat such entries as "not overweight" rather than propagating.
            float overweight = expectedDelta == 0f ? 0f : delta / expectedDelta * 100;
            if (!float.IsFinite(overweight))
            {
                overweight = 0f;
            }

            float percent = delta / totalDelta;
            // Calculate interest level
            int interest = Math.Abs(overweight) > 110 ? 1 : 0;
            interest += Math.Abs(percent) > 5 ? 1 : 0;
            interest += Math.Abs(percent) > 20 ? 1 : 0;
            interest += Math.Abs(percent) > 100 ? 1 : 0;
            interest += sourceValue / sourceTotal < 0.95 ? 1 : 0;  // Ignore top of the stack frames
            interest += sourceValue / sourceTotal < 0.75 ? 1 : 0;  // Bonus point for being further down the stack.

            results.Add(new OverWeightResult
            (
                Name: key,
                Before: baselineValue,
                After: sourceValue,
                Delta: delta,
                Overweight: overweight,
                Percent: percent,
                Interest: interest
            ));
        }

        results.Sort(static (left, right) =>
        {
            if (left.Interest < right.Interest)
                return 1;

            if (left.Interest > right.Interest)
                return -1;

            if (left.Overweight < right.Overweight)
                return 1;

            if (left.Overweight > right.Overweight)
                return -1;

            if (left.Delta < right.Delta)
                return -1;

            if (left.Delta > right.Delta)
                return 1;

            return 0;
        });

        return results.ToImmutable();
    }
```

4. **Test access.** `EtlDiffer` and `PerfDiff` are `internal`, and `tests/PerfDiff.Tests/PerfDiff.Tests.csproj` currently has **no reference to PerfDiff at all** (its 4 existing tests in `LoggingRegistrationTests.cs` exercise only the DI logging pattern). Add to `src/tools/PerfDiff/PerfDiff.csproj`:

```xml
  <ItemGroup>
    <InternalsVisibleTo Include="PerfDiff.Tests" />
  </ItemGroup>
```

   and to `tests/PerfDiff.Tests/PerfDiff.Tests.csproj`:

```xml
  <ItemGroup>
    <ProjectReference Include="..\..\src\tools\PerfDiff\PerfDiff.csproj" />
  </ItemGroup>
```

   (Skip either edit if a sibling PerfDiff correctness issue already added it.)

## Tests

New file `tests/PerfDiff.Tests/EtlDifferTests.cs` (xUnit arrives via `build/targets/tests/Tests.props`; no package edits needed). The `ListLogger` helper below is shared by the other PerfDiff correctness issues — add it once as `tests/PerfDiff.Tests/ListLogger.cs` if not already present:

```csharp
// tests/PerfDiff.Tests/ListLogger.cs
using Microsoft.Extensions.Logging;

namespace PerfDiff.Tests;

internal sealed class ListLogger : ILogger
{
    public List<(LogLevel Level, string Message)> Entries { get; } = [];

    public IDisposable? BeginScope<TState>(TState state)
        where TState : notnull
        => null;

    public bool IsEnabled(LogLevel logLevel) => true;

    public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception? exception, Func<TState, Exception?, string> formatter)
        => Entries.Add((logLevel, formatter(state, exception)));
}
```

```csharp
// tests/PerfDiff.Tests/EtlDifferTests.cs
using Microsoft.Extensions.Logging;
using PerfDiff.ETL;
using Xunit;

namespace PerfDiff.Tests;

public sealed class EtlDifferTests
{
    [Fact]
    public void TryCompareETL_CorruptZip_ReturnsFalseInsteadOfThrowing()
    {
        // Arrange: a file with the right extension but garbage content.
        string dir = Directory.CreateTempSubdirectory("perfdiff-etl-").FullName;
        try
        {
            string corrupt = Path.Combine(dir, "trace.etl.zip");
            File.WriteAllBytes(corrupt, new byte[] { 0x00, 0x01, 0x02, 0x03, 0xFF });
            ListLogger logger = new();

            // Act: must degrade gracefully (previously InvalidDataException escaped the catch filter).
            bool success = EtlDiffer.TryCompareETL(corrupt, corrupt, logger);

            // Assert
            Assert.False(success);
            Assert.Contains(logger.Entries, e => e.Level == LogLevel.Error && e.Message.Contains("ETL comparison failed", StringComparison.Ordinal));
        }
        finally
        {
            Directory.Delete(dir, recursive: true);
        }
    }

    [Fact]
    public void ComputeOverweights_ZeroBaselineTotal_ReturnsEmptyWithWarning()
    {
        ListLogger logger = new();
        Dictionary<string, float> sourceData = new() { ["Frame.A"] = 10f };
        Dictionary<string, float> baselineData = new() { ["Frame.A"] = 10f };

        var report = EtlDiffer.ComputeOverweights(100f, sourceData, 0f, baselineData, logger);

        Assert.Empty(report);
        Assert.Contains(logger.Entries, e => e.Level == LogLevel.Warning);
    }

    [Fact]
    public void ComputeOverweights_ZeroExpectedDelta_ProducesFiniteOverweight()
    {
        // baselineValue == 0 for one symbol -> expectedDelta == 0 -> previously Infinity/NaN overweight.
        ListLogger logger = new();
        Dictionary<string, float> sourceData = new() { ["Frame.A"] = 50f, ["Frame.B"] = 30f };
        Dictionary<string, float> baselineData = new() { ["Frame.A"] = 0f, ["Frame.B"] = 20f };

        var report = EtlDiffer.ComputeOverweights(200f, sourceData, 100f, baselineData, logger);

        Assert.All(report, r => Assert.True(float.IsFinite(r.Overweight), $"'{r.Name}' has non-finite Overweight {r.Overweight}"));
    }
}
```

Exit-code integrity test (the core B-1 behavior) in `tests/PerfDiff.Tests/PerfDiffCompareTests.cs` — a BDN regression plus present-but-unreadable ETL traces must still exit 1:

```csharp
// tests/PerfDiff.Tests/PerfDiffCompareTests.cs
using Xunit;

namespace PerfDiff.Tests;

public sealed class PerfDiffCompareTests
{
    private const string BdnJsonTemplate = """
        {
          "Title": "test",
          "Benchmarks": [
            {
              "FullName": "Suite.Bench1",
              "Statistics": { "Median": MEDIAN, "Mean": MEDIAN, "Percentiles": { "P95": MEDIAN } },
              "Measurements": [
                { "IterationStage": "Result", "Operations": 1, "Nanoseconds": V0 },
                { "IterationStage": "Result", "Operations": 1, "Nanoseconds": V1 },
                { "IterationStage": "Result", "Operations": 1, "Nanoseconds": V2 },
                { "IterationStage": "Result", "Operations": 1, "Nanoseconds": V3 },
                { "IterationStage": "Result", "Operations": 1, "Nanoseconds": V4 }
              ]
            }
          ]
        }
        """;

    private static string WriteRun(string root, string name, double[] valuesNs, double medianNs)
    {
        string dir = Path.Combine(root, name);
        Directory.CreateDirectory(dir);
        string json = BdnJsonTemplate
            .Replace("MEDIAN", medianNs.ToString("F1", System.Globalization.CultureInfo.InvariantCulture), StringComparison.Ordinal);
        for (int i = 0; i < valuesNs.Length; i++)
        {
            json = json.Replace($"V{i}", valuesNs[i].ToString("F1", System.Globalization.CultureInfo.InvariantCulture), StringComparison.Ordinal);
        }

        File.WriteAllText(Path.Combine(dir, "run-full-compressed.json"), json);
        return dir;
    }

    [Fact]
    public async Task CompareAsync_BdnRegressionWithUnreadableEtlPresent_ExitsOne()
    {
        string root = Directory.CreateTempSubdirectory("perfdiff-gate-").FullName;
        try
        {
            // Arrange: a clear 4x regression, plus etl.zip files that cannot be parsed.
            string baseline = WriteRun(root, "baseline", [100.0, 101, 102, 103, 104], 102.0);
            string results = WriteRun(root, "results", [400.0, 401, 402, 403, 404], 402.0);
            File.WriteAllBytes(Path.Combine(baseline, "trace.etl.zip"), new byte[] { 0x01, 0x02 });
            File.WriteAllBytes(Path.Combine(results, "trace.etl.zip"), new byte[] { 0x01, 0x02 });
            ListLogger logger = new();

            // Act: before the fix this either crashed (InvalidDataException) or, with valid
            // traces, returned 0 via the "noise" branch. ETL must not veto the BDN verdict.
            int exitCode = await PerfDiff.CompareAsync(baseline, results, failOnRegression: true, logger, CancellationToken.None);

            // Assert
            Assert.Equal(1, exitCode);
        }
        finally
        {
            Directory.Delete(root, recursive: true);
        }
    }
}
```

Note: if `TraceLog.OpenOrConvert` surfaces the corrupt-zip failure wrapped in a different exception type on this input (TraceEvent versions vary), the corrupt-zip test will show the actual type — extend the catch filter in `TryCompareETL` accordingly rather than weakening the test.

If issues #599/#600 (CLI and error-handling unit-test coverage) are implemented, the `CompareAsync` exit-code tests naturally belong in that suite; the `ComputeOverweights` tests belong with #600's edge-case coverage.

## Constraints

- **No new package dependencies.** `Directory.Packages.props` is pinned (Perfolizer 0.6.1, BenchmarkDotNet 0.15.8, TraceEvent as referenced). `InternalsVisibleTo`/`ProjectReference` are MSBuild items, not packages.
- Keep System.CommandLine 2.0.3 patterns in `Program.cs`/`DiffCommand.cs` untouched.
- All new diagnostics go through the existing `ILogger` abstractions (`SimpleConsoleLoggerProvider`); this issue removes the `Console.Error`/`Console.WriteLine` deviation in `EtlDiffer`.
- PerfDiff targets `net8.0` (`src/tools/PerfDiff/PerfDiff.csproj`); PerfDiff.Tests also targets `net8.0`.

## Validation

1. `dotnet build` — must complete with 0 warnings, 0 errors.
2. `dotnet test --settings ./build/targets/tests/test.runsettings` — all tests pass, including the 4 existing `LoggingRegistrationTests`.
3. `dotnet format` after changes; commit any formatting fixes.

## Acceptance criteria

- [ ] A BenchmarkDotNet-detected regression results in exit code 1 under `--failOnRegression` regardless of whether ETL traces are present, absent, or unreadable.
- [ ] `EtlDiffer.TryCompareETL` no longer exposes an `out bool regression` parameter (or, if the alternative is chosen, the verdict is genuinely derived from the report — not both).
- [ ] A corrupt `*.etl.zip` no longer crashes PerfDiff; it logs an error via `ILogger` and degrades gracefully.
- [ ] `ComputeOverweights` never emits `Infinity`/`NaN` `Overweight` values and handles `baselineTotal <= 0` and `sourceTotal <= 0` without dividing by zero.
- [ ] No `Console.WriteLine`/`Console.Error` calls remain in `EtlDiffer`.
- [ ] New unit tests above pass; build/test/format validation clean.

## Out of scope

- Deriving a calibrated ETL regression verdict (the alternative documented above) — requires real ETL trace data from CI to tune `Interest`/`Overweight` cuts.
- BDN benchmark-set coverage, threshold semantics, and ratio math — tracked separately as sibling correctness issues.
- Symbol-server configuration and `GetCallTree` behavior in `EtlDiffer` (unchanged).

## Related issues

- #599 / #600 — PerfDiff CLI and error-handling test coverage (natural home for the exit-code tests added here).
- #594 — PerfDiff core infrastructure tests.
- #609–#613 — PerfDiff tech-debt cleanups (this issue is a correctness defect, not tech debt, but touches the same files).
- ADR-008 (`docs/architecture/ADR-008-benchmarkdotnet-perfdiff-for-performance-regression-detection.md`) — defines the merge-gate contract this bug violates.
